<div style="color:#FFFFFF          ; display  :fill; border-radius:90px;
           background-color:#13A4B4; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :#FFFFFF; 
          text-align: Left;   font-size:28px; font-family  :cursive">       
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;Importing the Libraries<a id=5></a></p>
</div>

In [3]:
# import classes for Data Analysis
import numpy  as np
import pandas as pd

# import classes for data visulation
import matplotlib.pyplot    as plt
import seaborn              as sns
import plotly.express       as px
import plotly.graph_objects as go
import plotly.figure_factory as ff

import warnings
warnings.filterwarnings("ignore")

<div style="color:#FFFFFF          ; display  :fill; border-radius:90px;
           background-color:#13A4B4; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :#FFFFFF; 
          text-align: Left;   font-size:28px; font-family  :cursive">   
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp; Importing the Dataset <a id=6></a></p>

In [8]:
df = pd.read_csv('weatherHistory.csv')
# view dimensions of dataset
print("The Dimensions of Dataset is : ") 
df.shape

The Dimensions of Dataset is : 


(96453, 11)

In [10]:
#Previewing the Data set
df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Pressure (millibars)  \
0                     251          15.8263               1015.13   
1                     259          15.8263               1015.63   
2                     204          14.9569               1015.94   
3                     269          15.8263               1016.41   
4                     259          15.8263               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [11]:
df.tail()

Formatted Date        Summary Precip Type  \
96448  2016-09-09 19:00:00.000 +0200  Partly Cloudy        rain   
96449  2016-09-09 20:00:00.000 +0200  Partly Cloudy        rain   
96450  2016-09-09 21:00:00.000 +0200  Partly Cloudy        rain   
96451  2016-09-09 22:00:00.000 +0200  Partly Cloudy        rain   
96452  2016-09-09 23:00:00.000 +0200  Partly Cloudy        rain   

       Temperature (C)  Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
96448        26.016667                 26.016667      0.43            10.9963   
96449        24.583333                 24.583333      0.48            10.0947   
96450        22.038889                 22.038889      0.56             8.9838   
96451        21.522222                 21.522222      0.60            10.5294   
96452        20.438889                 20.438889      0.61             5.8765   

       Wind Bearing (degrees)  Visibility (km)  Pressure (millibars)  \
96448                      31          16.1000               1014.36   
96449                      20          15.5526               1015.16   
96450                      30          16.1000               1015.66   
96451                      20          16.1000               1015.95   
96452                      39          15.5204               1016.16   

                                Daily Summary  
96448  Partly cloudy starting in the morning.  
96449  Partly cloudy starting in the morning.  
96450  Partly cloudy starting in the morning.  
96451  Partly cloudy starting in the morning.  
96452  Partly cloudy starting in the morning.

# GETTING STARTED

<div style="color:#FFFFFF          ; display  :fill; border-radius:90px;
           background-color:#13A4B4; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :#FFFFFF; 
          text-align: Left;   font-size:28px; font-family  :cursive">          
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;&nbsp; Missing Value Detection  <a id=10></a>
</p>
</div>

In [13]:
# get the number of missing data points per column
missing_value_count = (df.isnull().sum())
print(missing_value_count[missing_value_count > 0])
# percentage of missing data
total_cells = np.product(df.shape)
total_missing_value = missing_value_count.sum()
print("Total percentage of our missing value is:",round((total_missing_value / total_cells * 100),4))
print('Total number of our cells is :',total_cells)
print('Total number of our missing value is :',total_missing_value)

Precip Type    517
dtype: int64
Total percentage of our missing value is: 0.0487
Total number of our cells is : 1060983
Total number of our missing value is : 517


<div style="color:#FFFFFF          ; display  :fill; border-radius:90px;
           background-color:#13A4B4; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :#FFFFFF; 
          text-align: Left;   font-size:28px; font-family  :cursive">          
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;&nbsp;Handling missing values  <a id=10></a>
</p>
</div>

In [14]:
df['Precip Type'].value_counts()

rain    85224
snow    10712
Name: Precip Type, dtype: int64

* As we can see that we have 517 missing values and unfortunately we can not **delete it**, because removing 517 index means removing 517 **random Formatted Date** which is can have **large negative impact** on our Analysis.
* As we can see that the majorty of Precip Type is rain so that why we will replace the missing values with rain variable.

In [15]:
# replace the missing values with rain variable
df['Precip Type'].fillna("rain", inplace = True)

**- So after filling the missing values, let's check it more again:**

In [16]:
df.isnull().sum()

Formatted Date              0
Summary                     0
Precip Type                 0
Temperature (C)             0
Apparent Temperature (C)    0
Humidity                    0
Wind Speed (km/h)           0
Wind Bearing (degrees)      0
Visibility (km)             0
Pressure (millibars)        0
Daily Summary               0
dtype: int64

**- So Now let's delete 'Daily Summary', we do not need it because we do have 'Summary' feature we will use that.**

In [17]:
df.drop(["Daily Summary"], axis=1, inplace=True)

**- Let's check it:**

In [18]:
df.columns

Index(['Formatted Date', 'Summary', 'Precip Type', 'Temperature (C)',
       'Apparent Temperature (C)', 'Humidity', 'Wind Speed (km/h)',
       'Wind Bearing (degrees)', 'Visibility (km)', 'Pressure (millibars)'],
      dtype='object')

# FEATURE ENGINEERING

<div style="color:#FFFFFF          ; display  :fill; border-radius:90px;
           background-color:#13A4B4; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :#FFFFFF; 
          text-align: Left;   font-size:28px; font-family  :cursive">          
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;&nbsp;Handling Dates  <a id=10></a>
</p>
</div>

In [20]:
#Changing Formatted Date from String to Datetime
df['Formatted Date'] = pd.to_datetime(df['Formatted Date'],utc=True)
df['Formatted Date'][0]

Timestamp('2006-03-31 22:00:00+0000', tz='UTC')

In [21]:
df = df.sort_values(by='Formatted Date')

<div style="color:#FFFFFF          ; display  :fill; border-radius:90px;
           background-color:#13A4B4; font-size:10px; font-family  :cursive">
    
<p style="padding   : .1px;    color    :#FFFFFF; 
          text-align: Left;   font-size:28px; font-family  :cursive">          
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;&nbsp;Create monthly and daily data frame  <a id=10></a>
</p>
</div>